Packet++/src/SomeIpSdLayer.cpp

```diff
@@ -0,0 +1,789 @@
+{
+	size_t lenEntries = getLenEntries();
+	int offsetToAddAt = sizeof(someipsdhdr) + sizeof(uint32_t) + lenEntries;
+
+	extendLayer(offsetToAddAt, entry.getLength());
+
+	setLenEntries(lenEntries + entry.getLength());
+
+	memcpy(m_Data + offsetToAddAt, entry.getDataPtr(), entry.getLength());
+
+	auto hdr = getSomeIpHeader();
+	hdr->length = htobe32(be32toh(hdr->length) + (uint32_t)entry.getLength());
+
+	return getNumEntries() - 1;
+}
+
+uint32_t SomeIpSdLayer::countOptions()
+{
+	size_t offsetOption = sizeof(someipsdhdr) + sizeof(uint32_t) + getLenEntries() + sizeof(uint32_t);
+	size_t lenOptions = getLenOptions();
+	uint32_t len = 0;
+
+	uint32_t numOptions = 0;
+	while (len < lenOptions)
+	{
+		uint32_t lenOption = be16toh(*((uint16_t *)(m_Data + offsetOption + len))) + 3 * sizeof(uint8_t);
+		len += lenOption;
+		++numOptions;
+	}
+	return numOptions;
+}
+
+uint32_t SomeIpSdLayer::findOption(const SomeIpSdOption &option)
+{
+	size_t offsetOption = sizeof(someipsdhdr) + sizeof(uint32_t) + getLenEntries() + sizeof(uint32_t);
+
+	uint32_t i = 0;
+	while (i < m_NumOptions)
+	{
+		uint32_t lenOption = be16toh(*((uint16_t *)(m_Data + offsetOption))) + 3 * sizeof(uint8_t);
+
+		if (option.getLength() == lenOption)
+		{
+			if (memcmp(m_Data + offsetOption, option.getDataPtr(), option.getLength()) == 0)
+			{
+				return i;
+			}
+		}
+
+		offsetOption += lenOption;
+		++i;
+	}
+	return i;
+}
+
+void SomeIpSdLayer::addOption(const SomeIpSdOption &option)
+{
+	int offsetToAddAt = (int)getHeaderLen();
+
+	extendLayer(offsetToAddAt, option.getLength());
+	memcpy(m_Data + offsetToAddAt, option.getDataPtr(), option.getLength());
+
+	setLenOptions(uint32_t(getLenOptions() + option.getLength()));
+
+	auto hdr = getSomeIpHeader();
+	hdr->length = htobe32(be32toh(hdr->length) + (uint32_t)option.getLength());
+
+	++m_NumOptions;
+}
+
+bool SomeIpSdLayer::addOptionIndex(uint32_t indexEntry, uint32_t indexOffset)
+{
+	/* The SOME/IP-SD protocol supports two option runs. Runs meaning that two different starting indices with differing
+	length can be provided. Of course, this only works if the indices in both runs are consecutive.
+
+		So, indices like this would work:
+			1 2 3 ; 7 8
+
+		What wouldn't work is this:
+			1 2 3 ; 7 9
+			1 3 ; 7 8
+	*/
+
+	size_t offsetToAddAt = sizeof(someipsdhdr) + sizeof(uint32_t) + indexEntry*sizeof(SomeIpSdEntry::someipsdhdrentry);
+	auto hdrEntry = (SomeIpSdEntry::someipsdhdrentry *)(m_Data + offsetToAddAt);
+
+	uint32_t indexFirstOption = (uint32_t)hdrEntry->indexFirstOption;
+	uint32_t lenFirstOption = (uint32_t)hdrEntry->nrOpt1;
+
+	if (lenFirstOption == 0)
+	{
+		hdrEntry->indexFirstOption = indexOffset;
+		++hdrEntry->nrOpt1;
+		return true;
+	}
+
+	if (indexFirstOption + lenFirstOption + 1 == indexOffset)
+	{
+		++hdrEntry->nrOpt1;
+		return true;
+	}
+
+	uint32_t indexSecondOption = (uint32_t)hdrEntry->indexSecondOption;
+	uint32_t lenSecondOption = (uint32_t)hdrEntry->nrOpt2;

```

Comment:
why casting to `uint32_t`? I think it'll be easier to stay with `uint8_t`, no?

Reply:
yes, not needed any more.